// @cucumber/cucumber/lib/runtime/coordinator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Coordinator = void 0;
const assemble_1 = require("../assemble");
/**
 * Handles the high-level coordination of the test run on the main thread
 * @returns whether all hooks and test cases were successful
 * @remarks
 * For each phase of the test run, delegates to a {@link RuntimeAdapter} to handle
 * the work itself. Unaware of how the adapter distributes and executes work.
 */
class Coordinator {
    testRunStartedId;
    eventBroadcaster;
    newId;
    sourcedPickles;
    supportCodeLibrary;
    adapter;
    constructor(testRunStartedId, eventBroadcaster, newId, sourcedPickles, supportCodeLibrary, adapter) {
        this.testRunStartedId = testRunStartedId;
        this.eventBroadcaster = eventBroadcaster;
        this.newId = newId;
        this.sourcedPickles = sourcedPickles;
        this.supportCodeLibrary = supportCodeLibrary;
        this.adapter = adapter;
    }
    async run() {
        try {
            await this.adapter.setup();
            const successByPhase = {
                beforeAllHooks: false,
                testCases: false,
                afterAllHooks: false,
            };
            successByPhase.beforeAllHooks = await this.adapter.runBeforeAllHooks();
            if (successByPhase.beforeAllHooks) {
                const assembledTestCases = await (0, assemble_1.assembleTestCases)(this.testRunStartedId, this.eventBroadcaster, this.newId, this.sourcedPickles, this.supportCodeLibrary);
                successByPhase.testCases = await this.adapter.runTestCases(assembledTestCases);
            }
            successByPhase.afterAllHooks = await this.adapter.runAfterAllHooks();
            return (successByPhase.beforeAllHooks && successByPhase.testCases && successByPhase.afterAllHooks);
        }
        finally {
            await this.adapter.teardown();
        }
    }
}
exports.Coordinator = Coordinator;
