// @cucumber/cucumber/lib/runtime/format_error.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.formatError = formatError;
const assertion_error_formatter_1 = require("assertion-error-formatter");
const error_stack_parser_1 = __importDefault(require("error-stack-parser"));
const filter_stack_trace_1 = require("../filter_stack_trace");
// Guards against runaway or self-referential cause chains.
const MAX_CAUSE_DEPTH = 10;
function formatError(error, filterStackTraces) {
    if (typeof error === 'string') {
        const stackTrace = `Error: ${error}`;
        return {
            message: stackTrace,
            exception: {
                type: 'Error',
                message: error,
                stackTrace,
            },
        };
    }
    const stackTrace = formatErrorAndCauses(error, filterStackTraces);
    const type = error.constructor.name;
    const message = typeof error === 'string' ? error : error.message;
    return {
        message: stackTrace,
        exception: {
            type,
            message,
            stackTrace,
        },
    };
}
function formatErrorAndCauses(error, filterStackTraces) {
    let output = formatSingleError(error, filterStackTraces);
    const seen = new Set([error]);
    let cause = error.cause;
    for (let depth = 0; cause !== undefined && cause !== null; depth++) {
        if (depth >= MAX_CAUSE_DEPTH) {
            output += '\nCaused by: ... (further causes truncated)';
            break;
        }
        if (seen.has(cause)) {
            output += '\nCaused by: ... (circular reference)';
            break;
        }
        seen.add(cause);
        if (cause instanceof Error) {
            output += `\nCaused by: ${formatSingleError(cause, filterStackTraces)}`;
            cause = cause.cause;
        }
        else {
            output += `\nCaused by: ${String(cause)}`;
            break;
        }
    }
    return output;
}
function formatSingleError(error, filterStackTraces) {
    let processedStackTrace = error.stack;
    try {
        const parsedStack = error_stack_parser_1.default.parse(error);
        const filteredStack = filterStackTraces ? (0, filter_stack_trace_1.filterStackTrace)(parsedStack) : parsedStack;
        processedStackTrace = filteredStack.map((f) => f.source).join('\n');
    }
    catch {
        // if we weren't able to parse and process, we'll settle for the original
    }
    return (0, assertion_error_formatter_1.format)(error, {
        colorFns: {
            errorStack: (stack) => {
                return processedStackTrace ? `\n${processedStackTrace}` : stack;
            },
        },
    });
}
