// @cucumber/cucumber/lib/cli/run.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = run;
const _1 = __importDefault(require("./"));
const validate_node_engine_version_1 = require("./validate_node_engine_version");
function logErrorMessageAndExit(message) {
    // biome-ignore lint/suspicious/noConsole: cli entrypoint, other code abstracts console for testability
    console.error(message);
    process.exit(1);
}
async function run() {
    (0, validate_node_engine_version_1.validateNodeEngineVersion)(process.version, (error) => {
        // biome-ignore lint/suspicious/noConsole: cli entrypoint, other code abstracts console for testability
        console.error(error);
        process.exit(1);
    }, 
    // biome-ignore lint/suspicious/noConsole: cli entrypoint, other code abstracts console for testability
    console.warn);
    const cli = new _1.default({
        argv: process.argv,
        cwd: process.cwd(),
        stdout: process.stdout,
        stderr: process.stderr,
        env: process.env,
    });
    let result;
    try {
        result = await cli.run();
    }
    catch (error) {
        logErrorMessageAndExit(error);
    }
    const exitCode = result.success ? 0 : 1;
    if (result.shouldExitImmediately) {
        process.exit(exitCode);
    }
    else {
        process.exitCode = exitCode;
    }
}
