// @cucumber/cucumber/lib/configuration/check_schema.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.checkSchema = checkSchema;
const gherkin_1 = require("@cucumber/gherkin");
const yup = __importStar(require("yup"));
const schema = yup.object().shape({
    backtrace: yup.boolean(),
    dryRun: yup.boolean(),
    exit: yup.boolean(),
    failFast: yup.boolean(),
    format: yup
        .array()
        .of(yup.lazy((val) => Array.isArray(val) ? yup.array().of(yup.string()).min(1).max(2) : yup.string())),
    formatOptions: yup
        .object({
        colorsEnabled: yup.boolean(),
        html: yup.object(),
        includeAttachments: yup.boolean(),
        pretty: yup.object(),
        printAttachments: yup.boolean(),
        rerun: yup.object(),
        theme: yup.object(),
    })
        .noUnknown(false),
    import: yup.array().of(yup.string()),
    language: yup.string().oneOf(Object.keys(gherkin_1.dialects)),
    name: yup.array().of(yup.string()),
    order: yup.string().matches(/^random:.*|random|defined$/),
    paths: yup.array().of(yup.string()),
    parallel: yup.number().integer().min(0),
    plugin: yup.array().of(yup.string().trim().required()),
    pluginOptions: yup.object(),
    publish: yup.boolean(),
    require: yup.array().of(yup.string()),
    requireModule: yup.array().of(yup.string()),
    retry: yup.number().integer().min(0),
    retryTagFilter: yup.string(),
    strict: yup.boolean(),
    tags: yup.string(),
    workerOptions: yup.object({
        resourceLimits: yup.object(),
    }),
    worldParameters: yup.object(),
});
function checkSchema(configuration) {
    return schema.validateSync(configuration, {
        abortEarly: false,
        strict: true,
        stripUnknown: true,
    });
}
