// @cucumber/cucumber/lib/runtime/parallel/test_cases_phase.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestCasesPhase = void 0;
class TestCasesPhase {
    resolve;
    reject;
    logger;
    canAssign;
    failing = false;
    idleInterventions = 0;
    queue = [];
    running = new Set();
    constructor(resolve, reject, logger, canAssign, assembledTestCases) {
        this.resolve = resolve;
        this.reject = reject;
        this.logger = logger;
        this.canAssign = canAssign;
        this.queue.push(...assembledTestCases);
    }
    fill() {
        return this.select();
    }
    next(command, event) {
        if (!event.success) {
            this.failing = true;
        }
        this.running.delete(command.assembledTestCase.pickle);
        if (this.queue.length === 0 && this.running.size === 0) {
            if (this.idleInterventions > 0) {
                this.logger.warn(`WARNING: All workers went idle ${this.idleInterventions} time(s). Consider revising handler passed to setParallelCanAssign.`);
            }
            this.resolve(!this.failing);
            return undefined;
        }
        return this.select();
    }
    select() {
        if (this.queue.length === 0) {
            return undefined;
        }
        for (const assembledTestCase of this.queue) {
            if (this.canAssign(assembledTestCase.pickle, [...this.running])) {
                return this.dequeue(assembledTestCase);
            }
        }
        if (this.running.size === 0) {
            this.idleInterventions++;
            return this.dequeue(this.queue.at(0));
        }
        return undefined;
    }
    dequeue(assembledTestCase) {
        this.queue.splice(this.queue.indexOf(assembledTestCase), 1);
        this.running.add(assembledTestCase.pickle);
        return {
            type: 'TEST_CASE',
            assembledTestCase,
            failing: this.failing,
        };
    }
}
exports.TestCasesPhase = TestCasesPhase;
